python/example_code/s3/s3_versioning/versioning.py

```diff
@@ -0,0 +1,259 @@
+    that expires noncurrent object versions after 7 days.
+
+    Adding a lifecycle configuration to a versioned bucket is a best practice.
+    It helps prevent objects in the bucket from accumulating a large number of
+    noncurrent versions, which can slow down request performance.
+
+    Usage is shown in the usage_demo_single_object function at the end of this module.
+
+    :param bucket_name: The name of the bucket to create.
+    :return: The newly created bucket.
+    """
+    try:
+        bucket = s3.create_bucket(
+            Bucket=bucket_name,
+            CreateBucketConfiguration={
+                'LocationConstraint': s3.meta.client.meta.region_name
+            }
+        )
+        logger.info("Created bucket %s.", bucket.name)
+    except ClientError as error:
+        if error.response['Error']['Code'] == 'BucketAlreadyOwnedByYou':
+            logger.warning("Bucket %s already exists! Using it.", bucket_name)
+            bucket = s3.Bucket(bucket_name)
+        else:
+            logger.exception("Couldn't create bucket %s.", bucket_name)
+            raise
+
+    try:
+        bucket.Versioning().enable()
+        logger.info("Enabled versioning on bucket %s.", bucket.name)
+    except ClientError:
+        logger.exception("Couldn't enable versioning on bucket %s.", bucket.name)
+        raise
+
+    try:
+        expiration = 7
+        bucket.LifecycleConfiguration().put(
+            LifecycleConfiguration={
+                'Rules': [{
+                    'Status': 'Enabled',
+                    'Prefix': prefix,
+                    'NoncurrentVersionExpiration': {'NoncurrentDays': expiration}
+                }]
+            }
+        )
+        logger.info("Configured lifecycle to expire noncurrent versions after %s days "
+                    "on bucket %s.", expiration, bucket.name)
+    except ClientError as error:
+        logger.warning("Couldn't configure lifecycle on bucket %s because %s. "
+                       "Continuing anyway.", bucket.name, error)
+
+    return bucket
+
+
+def rollback_object(bucket, object_key, version_id):
+    """
+    Rolls back an object to an earlier version by deleting all versions that
+    occurred after the specified rollback version.
+
+    Usage is shown in the usage_demo_single_object function at the end of this module.
+
+    :param bucket: The bucket that holds the object to roll back.
+    :param object_key: The object to roll back.
+    :param version_id: The version ID to roll back to.
+    """
+    # Versions must be sorted by last_modified date because delete markers are
+    # at the end of the list even when they are interspersed in time.
+    versions = sorted(bucket.object_versions.filter(Prefix=object_key),
+                      key=attrgetter('last_modified'), reverse=True)
+
+    logger.debug(
+        "Got versions:\n%s",
+        '\n'.join([f"    {version.version_id}, last modified {version.last_modified}"
```

Comment:
Yes, based on date of release and features, Python 3.6 is the current sweet spot for examples. If you're using an older version of Python you can probably figure out how to retrofit fairly easily. If we get feedback that says otherwise, we can revisit.